File: src/synapse_agent/tools/jina_search.py
"""Jina Search API — primary search provider."""

from __future__ import annotations

import httpx

from synapse_agent.config import settings
from synapse_agent.state import SearchResult


async def jina_search(query: str, num_results: int = 5) -> list[SearchResult]:
    """Search via Jina Search API. Returns structured results."""
    if not settings.jina_api_key:
        return []

    headers = {
        "Authorization": f"Bearer {settings.jina_api_key}",
        "Accept": "application/json",
        "X-Retain-Images": "none",
    }

    async with httpx.AsyncClient(timeout=30) as client:
        try:
            resp = await client.get(
                f"https://s.jina.ai/{query}",
                headers=headers,
            )
            resp.raise_for_status()
            data = resp.json()
        except Exception:
            return []

    results = []
    for item in data.get("data", [])[:num_results]:
        results.append(SearchResult(
            url=item.get("url", ""),
            title=item.get("title", ""),
            snippet=item.get("description", item.get("content", "")[:500]),
            content=item.get("content", ""),
            source_query=query,
        ))
    return results

File: src/lib/synapse/tools/jina-search.ts
/**
 * Jina Search API — primary search provider.
 * Port of Python's tools/jina_search.py
 */

import type { SearchResult } from "../types";
import { getSettings } from "../config";

export async function jinaSearch(
  query: string,
  numResults = 5
): Promise<SearchResult[]> {
  const { jinaApiKey } = getSettings();
  if (!jinaApiKey) return [];

  try {
    const resp = await fetch(`https://s.jina.ai/${encodeURIComponent(query)}`, {
      headers: {
        Authorization: `Bearer ${jinaApiKey}`,
        Accept: "application/json",
        "X-Retain-Images": "none",
      },
      signal: AbortSignal.timeout(30_000),
    });

    if (!resp.ok) return [];
    const data = await resp.json();

    return (data.data ?? [])
      .slice(0, numResults)
      .map(
        (item: Record<string, string>): SearchResult => ({
          url: item.url ?? "",
          title: item.title ?? "",
          snippet:
            item.description ?? (item.content ?? "").slice(0, 500),
          content: item.content ?? "",
          relevanceScore: 0,
          sourceQuery: query,
        })
      );
  } catch {
    return [];
  }
}
